### PR TITLE
Add download to release files

### PR DESCRIPTION
**Pull request by HazAT**
kapture 2016-12-21 at 16 24 35

### CODE REVIEW

**Review by mitsuhiko — APPROVED** [review]

**Comment by mitsuhiko** [comment]
Looks good to me. Might need a CHANGES entry

**Comment by mattrobenolt** [comment]
No no, we explicitly cannot do this for security reasons. There are specific reasons why uploads are a one-way.

**Review by mattrobenolt — CHANGES_REQUESTED** [review]
We need to surface this ability possibly only for superusers only or in our admin.

**Comment by mitsuhiko** [comment]
Can we put this behind a org admin permission / superuser flag or something
like thatat least? It's super useful for support on react native

**Comment by mattrobenolt** [comment]
Unsure what we wanted to do, but @dcramer and @benvinegar should weigh their thoughts. 

I personally think behind superuser check is fine.

**Comment by dcramer** [comment]
What's the security concern?

On Dec 21, 2016 7:52 AM, "Matt Robenolt" <notifications@github.com> wrote:

Unsure what we wanted to do, but @dcramer <https://github.com/dcramer> and
@benvinegar <https://github.com/benvinegar> should weigh their thoughts.

I personally think behind superuser check is fine.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/getsentry/sentry/pull/4704#issuecomment-268557243>, or mute
the thread

**Comment by mattrobenolt** [comment]
> What's the security concern?

We never wanted having access to downloading release artifacts through the UI since this would expose source code to people on the org who maybe didn't have access before.

**Comment by dcramer** [comment]
@mattrobenolt then we scope it to the same perms. They already have indirect access via source expansion.

**Comment by mitsuhiko** [comment]
@dcramer by same perms you mean `project:releases`?

**Comment by dcramer** [comment]
Yeah I think so -- will defer to others here, but its my feeling that these aren't highly sensitive, and should be accessible by the same users who have permission to create/delete them.

**Comment by mitsuhiko** [comment]
I would just say we add an explicit check for `projects:releases` or `projects:write` and if one of them passes we permit the download.

File: src/sentry/api/endpoints/release_file_details.py
Comment (mattrobenolt) [inline src/sentry/api/endpoints/release_file_details.py]: We should do the same thing that Django does here for static file delivery, since it's the same idea: https://github.com/django/django/blob/1.6.11/django/views/static.py#L67-L68

The reason for this is if it's this `CompatibleStreamingHttpResponse`, we can actually stream the bytes back to the browser instead of blocking process while it downloads in memory. This streaming behavior is then also handled directly by uWSGI and deferred away from Django.

This happens here: https://github.com/getsentry/sentry/blob/master/src/sentry/wsgi.py#L44

Comment (mattrobenolt) [inline src/sentry/api/endpoints/release_file_details.py]: We should also extract the `mimetype` from `file.headers.get('content-type', 'application/octet-stream')` since we typically have this information.

File: src/sentry/api/endpoints/release_file_details.py
Comment (mattrobenolt) [inline src/sentry/api/endpoints/release_file_details.py]: We should stick with just `project:write`.

Checking for `project:releases` is kinda moot at this point anyways since literally any member has `project:releases` permission. But as of now, `project:releases` doesn't have any concept of `read` vs `write`, so this is effectively opening it up to everybody.

We should limit to `project:write` and document this behavior. We can always loosen constraint here if we have the need in the future. I'm hesitant on exposing this feature to people who potentially didn't have access to source code before, so I feel `project:write` is sufficient to start.

File: src/sentry/static/sentry/app/views/releaseArtifacts.jsx
Comment (mattrobenolt) [inline src/sentry/static/sentry/app/views/releaseArtifacts.jsx]: Same here for `project:write`.

According to @dcramer, which I agree with, we should leave the button here even if they don't have permission, but disable it with a tooltip saying that they don't have permission. That way we're not hiding things in the UI and they can see easily that they don't have permission to perform an action.

**Review by mattrobenolt — CHANGES_REQUESTED** [review]

File: src/sentry/api/endpoints/release_file_details.py
Comment (mattrobenolt) [inline src/sentry/api/endpoints/release_file_details.py]: To be more explicit, this last line could be:

```python
CompatibleStreamingHttpResponse(
    iter(lambda: file.read(4096), b''),
    content_type=file.headers.get('content-type', 'application/octet-stream'),
)
```

Comment (mitsuhiko) [inline src/sentry/api/endpoints/release_file_details.py]: Oh I see. I thought `HttpResponse` already can handle this on its own but this just works by chance. Clever :P

**Comment by mitsuhiko** [comment]
Just needs changes and a sign-off from @mattrobenolt i think.

**Review by mattrobenolt — CHANGES_REQUESTED** [review]
You need to revert changes to the locale files.

File: src/sentry/api/endpoints/release_file_details.py
Comment (mattrobenolt) [inline src/sentry/api/endpoints/release_file_details.py]: We should also probably return a 403 here when requesting a download but don't have have permission.

**Comment by HazAT** [comment]
@mattrobenolt is there anything left to do?

**Comment by mattrobenolt** [comment]
We need some tests for this. They can be pretty trivial.

**Comment by HazAT** [comment]
Also wanted to write a api integration test but this
https://github.com/getsentry/sentry/pull/4704/files#diff-17bacf0f00a6813b38ad91e1b0efa99eR76
raises a `raise ValueError('Cannot seek to pos')`
Don't know how to have a seekable file for testing...

**Comment by mitsuhiko** [comment]
cStringIO.StringIO can seek over binary data. Not sure if that helps.

**Comment by mattrobenolt** [comment]
Can you commit the test code that fails? Hard to debug without seeing what you're doing. :)

File: tests/sentry/api/endpoints/test_release_file_details.py
Comment (mitsuhiko) [inline tests/sentry/api/endpoints/test_release_file_details.py]: I think you need to create the file before the release file and then use `putfile` on it:

```python
from io import BytesIO
f = File.objects.create(...)
f.putfile(BytesIO('File contents here'))
ReleaseFIle.objects.create(..., file=f)
```

**Review by mitsuhiko — CHANGES_REQUESTED** [review]

**Comment by mattrobenolt** [comment]
Yeah, that's it. :)

--
Matt Robenolt
@mattrobenolt




On Sun, Jan 1, 2017 at 11:59 AM -0800, "Armin Ronacher" <notifications@github.com> wrote:












@mitsuhiko requested changes on this pull request.





In tests/sentry/api/endpoints/test_release_file_details.py:
> +        self.login_as(user=self.user)
+
+        project = self.create_project(name='foo')
+
+        release = Release.objects.create(
+            project=project,
+            organization_id=project.organization_id,
+            version='1',
+        )
+        release.add_project(project)
+
+        releasefile = ReleaseFile.objects.create(
+            organization_id=project.organization_id,
+            project=project,
+            release=release,
+            file=File.objects.create(



I think you need to create the file before the release file and then use putfile on it:
from io import BytesIO
f = File.objects.create(...)
f.putfile(BytesIO('File contents here'))
ReleaseFIle.objects.create(..., file=f)

**Comment by HazAT** [comment]
Thx guys :)

**Comment by mitsuhiko** [comment]
Now only needs a CHANGES entry and this is ready to go.

**Review by mitsuhiko — APPROVED** [review]

File: tests/sentry/api/endpoints/test_release_file_details.py
Comment (mattrobenolt) [inline tests/sentry/api/endpoints/test_release_file_details.py]: It'd be worth adding an assertion for the contents of the payload here to match the `BytesIO` contents to confirm that we've actually downloaded the correct file.

As well as the correct headers. `Content-Length`, `Content-Type`, and `Content-Disposition`. Mostly because in this case, these things are actually significant and we wouldn't want to break them.

Other than this, lgtm.

**Review by mattrobenolt — APPROVED** [review]

**Comment by HazAT** [comment]
Squash Merge?

**Comment by mitsuhiko** [comment]
+1

**Pull request merged**
